File: Checklists_Code/Checklists_Code/ListDetailViewController.m
//
//  ListDetailViewController.m
//  Checklists_Code
//
//  Created by wangx on 15/6/24.
//  Copyright (c) 2015年 wxiang1991. All rights reserved.
//

#import "ListDetailViewController.h"
#import "IconPickerViewController.h"
#import "Checklist.h"
@interface ListDetailViewController (){
    NSString *_name;
    NSString *_iconName;
    
}

@end

@implementation ListDetailViewController

-(void)loadView{
    [super loadView];
    UIBarButtonItem *cancelBarButtonItem=[[UIBarButtonItem alloc]initWithBarButtonSystemItem:UIBarButtonSystemItemCancel target:self action:@selector(cancel)];
    self.navigationItem.leftBarButtonItem=cancelBarButtonItem;
    
    self.doneBarButton=[[UIBarButtonItem alloc]initWithBarButtonSystemItem:UIBarButtonSystemItemDone target:self action:@selector(done)];
    self.navigationItem.rightBarButtonItem=self.doneBarButton;
}
- (void)viewDidLoad {
    [super viewDidLoad];
    if (self.checklistToEdit!=nil) {
        self.title=@"Edit CheckList";
        _name=self.checklistToEdit.name;
        _iconName=self.checklistToEdit.iconName;
    }else{
        self.title=@"Add CheckList";
        _iconName = @"Folder";
    }
    self.iconImageView.image=[UIImage imageNamed:_iconName];
    self.textField.delegate=self;
    [self.textField becomeFirstResponder];
    
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return 2;
}


- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    static NSString *cellIdentifier=@"Cell";
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:cellIdentifier];
    if (cell==nil) {
        cell=[[UITableViewCell alloc]initWithStyle:UITableViewCellStyleDefault reuseIdentifier:cellIdentifier];
    }
    if (indexPath.row==0) {
        self.textField=[[UITextField alloc]initWithFrame:CGRectMake(10, 7, 300, 30)];
        self.textField.placeholder=@"Name of List";
        self.textField.text=_name;
        [cell.contentView addSubview:self.textField];
    }else{
        self.label=[[UILabel alloc]initWithFrame:CGRectMake(20, 11, 150, 20)];
        self.label.text=_iconName;
        self.label.textAlignment=NSTextAlignmentCenter;
        [cell.contentView addSubview:self.label];
        
        self.iconImageView=[[UIImageView alloc]initWithFrame:CGRectMake(200, 4, 36, 36)];
        self.iconImageView.image=[UIImage imageNamed:_iconName];
        cell.accessoryType=UITableViewCellAccessoryDisclosureIndicator;
        [cell.contentView addSubview:self.iconImageView];
    }
    return cell;
}
//点击取消按钮时，触发事件
-(void)cancel{
    //调用代理方法，回弹到上一个页面
    [self.delegate listDetailViewControllerDidCancel:self];
}
//点击done按钮时，触发事件
-(void)done{
    //如果传过来的checklistToEdit==nil 则增加，否则修改
    if (self.checklistToEdit==nil) {
        Checklist *checklist=[[Checklist alloc]init];
        checklist.name=[NSString stringWithString:self.textField.text];
        checklist.iconName=_iconName;
        [self.delegate listDetailViewController:self didFinishAddingChecklist:checklist];
    }else{
        self.checklistToEdit.name=self.textField.text;
        self.checklistToEdit.iconName=_iconName;
        [self.delegate listDetailViewController:self didFinishEditingChecklist:self.checklistToEdit];
    }
}
#pragma mark 点击第二行的时候进入选择icon的功能
-(void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath{
    if (indexPath.row==1) {
        IconPickerViewController *controller=[[IconPickerViewController alloc]init];
        controller.delegate=self;
        [self.navigationController pushViewController:controller animated:YES];
    }
}

#pragma mark UITextFieldDelegate
-(BOOL)textField:(UITextField *)textField shouldChangeCharactersInRange:(NSRange)range replacementString:(NSString *)string{
    NSString *newText = [textField.text stringByReplacingCharactersInRange:range withString:string];
    self.doneBarButton.enabled = ([newText length] > 0);
    return YES;
}

#pragma mark IconPickerViewControllerDelegate
-(void)iconPicker:(IconPickerViewController *)picker didPickIcon:(NSString *)iconName{
    _iconName=iconName;
    self.iconImageView.image=[UIImage imageNamed:_iconName];
    self.label.text=_iconName;
    [self.navigationController popToViewController:self animated:YES];
}


/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: ChecklistSwift/ChecklistSwift/ListDetailViewController.swift
//
//  ListDetailViewController.swift
//  ChecklistSwift
//
//  Created by 王享 on 15/12/28.
//  Copyright © 2015年 王享. All rights reserved.
//

import UIKit
protocol ListDetailViewControllerDelegate {
    
    func listDetailViewControllerDidCancel(_ controller:ListDetailViewController)
    
    func listDetailViewController(_ controller:ListDetailViewController,didFinishAddingChecklist checklist:Checklist)
    
    func listDetailViewController(_ controller:ListDetailViewController,didFinishEditingChecklist checklist:Checklist)
    
}


class ListDetailViewController: UITableViewController,UITextFieldDelegate,IconPickerViewControllerDelegate {

    var textField:UITextField?
    var label:UILabel?
    var iconImageView:UIImageView?
    var doneBarButton:UIBarButtonItem?
    var checklistToEdit:Checklist?
    var delegate:ListDetailViewControllerDelegate?
    
    var _name:String?
    var _iconName:String?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        let cancelBarButtonItem:UIBarButtonItem=UIBarButtonItem.init(barButtonSystemItem: UIBarButtonSystemItem.cancel, target: self, action: #selector(ListDetailViewController.cancel))
        self.navigationItem.leftBarButtonItem = cancelBarButtonItem
        
        self.doneBarButton = UIBarButtonItem.init(barButtonSystemItem: UIBarButtonSystemItem.done, target: self, action: #selector(ListDetailViewController.done));
        self.navigationItem.rightBarButtonItem = self.doneBarButton
        
        if(self.checklistToEdit != nil){
            self.title = "Edit CheckList"
            _name = self.checklistToEdit?.name
            _iconName = self.checklistToEdit?.iconName
        }else{
            self.title = "Add CheckList"
            _iconName = "Folder"
        }
        self.iconImageView?.image = UIImage.init(named: _iconName!)
        self.textField?.delegate = self
        self.textField?.becomeFirstResponder()
    }
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return 2;
    }
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let idetidier:String = "Cell"
        var cell:UITableViewCell? = tableView.dequeueReusableCell(withIdentifier: idetidier)
        if(cell == nil ){
            cell = UITableViewCell.init(style: UITableViewCellStyle.default, reuseIdentifier: idetidier)
        }
        if indexPath.row == 0 {
            self.textField = UITextField.init(frame: CGRect(x: 10, y: 7, width: 300, height: 30))
            self.textField?.placeholder="Name of List"
            self.textField?.text = _name
            cell?.contentView.addSubview(self.textField!)
        }else{
            self.label = UILabel.init(frame:CGRect(x: 20, y: 11, width: 150, height: 20))
            self.label?.text = _iconName
            self.label?.textAlignment = NSTextAlignment.center
            cell?.contentView.addSubview(self.label!)
            
            self.iconImageView = UIImageView.init(frame: CGRect(x: 200, y: 4, width: 36, height: 36))
            if(_iconName != nil){
                self.iconImageView?.image=UIImage.init(named:_iconName!)
            }
            cell?.accessoryType = UITableViewCellAccessoryType.disclosureIndicator
            [cell?.contentView .addSubview(self.iconImageView!)]
        }
        return cell!
    }
    override func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        if(indexPath.section==0&&indexPath.row==3){
            return 220.0
        }else{
            return UITableViewAutomaticDimension
        }
    }
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        if(indexPath.row == 1){
            let controller:IconPickerViewController = IconPickerViewController()
            controller.delegate=self
            self.navigationController?.pushViewController(controller, animated: true)
        }
    }
    

    func cancel(){
        self.delegate?.listDetailViewControllerDidCancel(self)
    }
    
    func done(){
        if(self.checklistToEdit == nil){
            checklistToEdit = Checklist.init();
            checklistToEdit?.name=self.textField?.text
            checklistToEdit?.iconName=_iconName
            self.delegate?.listDetailViewController(self, didFinishAddingChecklist: checklistToEdit!)
        }else{
            self.checklistToEdit?.name=self.textField?.text
            self.checklistToEdit?.iconName=_iconName
            self.delegate?.listDetailViewController(self, didFinishEditingChecklist: self.checklistToEdit!);
        }
    }
    func textField(_ textField: UITextField, shouldChangeCharactersIn range: NSRange, replacementString string: String) -> Bool {
        let newText:String = textField.text!;
        self.doneBarButton?.isEnabled = !newText.isEmpty
        return true
    }
    
    func iconPicker(_ picker:IconPickerViewController, didPickIcon iconName:String){
        _iconName = iconName;
        self.iconImageView!.image = UIImage.init(named: _iconName!)
        self.label!.text = _iconName;
        self.navigationController!.popViewController(animated: true)
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }
}
